mypy/checkexpr.py

```diff
@@ -1350,91 +1347,62 @@ def overload_erased_call_targets(self,
-        """Accepts a list of overload signatures and attempts to combine them together into a
-        new CallableType consisting of the union of all of the given arguments and return types.
-
-        Returns None if it is not possible to combine the different callables together in a
-        sound manner.
-
-        Assumes all of the given callables have argument counts compatible with the caller.
+    def union_overload_result(self,
+                              plausible_targets: List[CallableType],
+                              args: List[Expression],
+                              arg_types: List[Type],
+                              arg_kinds: List[int],
+                              arg_names: Optional[Sequence[Optional[str]]],
+                              callable_name: Optional[str],
+                              object_type: Optional[Type],
+                              context: Context,
+                              arg_messages: Optional[MessageBuilder] = None,
+                              ) -> Optional[Tuple[Type, Type]]:
+        """Accepts a list of overload signatures and attempts to match calls by destructuring
+        the first union. Returns None if there is no match.
         """
-        if len(callables) == 0:
-            return None
-        elif len(callables) == 1:
-            return callables[0]
-
-        # Note: we are assuming here that if a user uses some TypeVar 'T' in
-        # two different overloads, they meant for that TypeVar to mean the
-        # same thing.
-        #
-        # This function will make sure that all instances of that TypeVar 'T'
-        # refer to the same underlying TypeVarType and TypeVarDef objects to
-        # simplify the union-ing logic below.
-        #
-        # (If the user did *not* mean for 'T' to be consistently bound to the
-        # same type in their overloads, well, their code is probably too
-        # confusing and ought to be re-written anyways.)
-        callables, variables = merge_typevars_in_callables_by_name(callables)
-
-        new_args = [[] for _ in range(len(callables[0].arg_types))]  # type: List[List[Type]]
-        new_kinds = list(callables[0].arg_kinds)
-        new_returns = []  # type: List[Type]
-
-        for target in callables:
-            # We conservatively end if the overloads do not have the exact same signature.
-            # The only exception is if one arg is optional and the other is positional: in that
-            # case, we continue unioning (and expect a positional arg).
-            # TODO: Enhance the union overload logic to handle a wider variety of signatures.
-            if len(new_kinds) != len(target.arg_kinds):
+        if not any(self.real_union(typ) for typ in arg_types):
+            # No unions in args, just fall back to normal inference
+            for arg, typ in zip(args, arg_types):
+                self.type_overrides[arg] = typ
```

Comment:
This should be made a context manager, also we need to be sure that the last type for an arg expression (the one that stays in type map after we leave the call expression) is a union.